File: apps/api/src/main/java/br/unifor/costify/application/usecase/CalculateRecipeCostUseCase.java
package br.unifor.costify.application.usecase;

import br.unifor.costify.application.contracts.RecipeRepository;
import br.unifor.costify.application.dto.response.RecipeCostDto;
import br.unifor.costify.application.errors.RecipeNotFoundException;
import br.unifor.costify.application.service.IngredientLoaderService;
import br.unifor.costify.domain.entity.Ingredient;
import br.unifor.costify.domain.entity.Recipe;
import br.unifor.costify.domain.service.RecipeCostCalculationService;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.RecipeCost;
import br.unifor.costify.domain.valueobject.RecipeIngredient;

import java.util.Map;
import java.util.Optional;

public class CalculateRecipeCostUseCase {
    
    private final RecipeRepository recipeRepository;
    private final IngredientLoaderService ingredientLoaderService;
    private final RecipeCostCalculationService costCalculationService;
    
    public CalculateRecipeCostUseCase(
            RecipeRepository recipeRepository,
            IngredientLoaderService ingredientLoaderService,
            RecipeCostCalculationService costCalculationService) {
        this.recipeRepository = recipeRepository;
        this.ingredientLoaderService = ingredientLoaderService;
        this.costCalculationService = costCalculationService;
    }
    
    public RecipeCostDto execute(String recipeId) {
        Id id = Id.of(recipeId);
        
        Optional<Recipe> recipeOpt = recipeRepository.findById(id);
        if (recipeOpt.isEmpty()) {
            throw new RecipeNotFoundException("Recipe not found with ID: " + recipeId);
        }
        
        Recipe recipe = recipeOpt.get();
        Map<Id, Ingredient> ingredientMap = ingredientLoaderService.loadIngredients(recipe.getIngredients());
        
        RecipeCost recipeCost = costCalculationService.calculateCost(recipe, ingredientMap);
        
        return RecipeCostDto.fromDomain(recipeCost);
    }
}

File: internal/application/usecase/calculate_recipe_cost_usecase.go
package usecase

import (
	"github.com/vini/costify-go/internal/application/contract"
	"github.com/vini/costify-go/internal/application/dto/response"
	"github.com/vini/costify-go/internal/application/errors"
	"github.com/vini/costify-go/internal/application/service"
	domainservice "github.com/vini/costify-go/internal/domain/service"
	"github.com/vini/costify-go/internal/domain/valueobject"
)

// CalculateRecipeCostUseCase handles recipe cost calculation
type CalculateRecipeCostUseCase struct {
	recipeRepository        contract.RecipeRepository
	ingredientLoaderService *service.IngredientLoaderService
	costCalculationService  *domainservice.RecipeCostCalculationService
}

// NewCalculateRecipeCostUseCase creates a new CalculateRecipeCostUseCase
func NewCalculateRecipeCostUseCase(
	recipeRepository contract.RecipeRepository,
	ingredientLoaderService *service.IngredientLoaderService,
	costCalculationService *domainservice.RecipeCostCalculationService,
) *CalculateRecipeCostUseCase {
	return &CalculateRecipeCostUseCase{
		recipeRepository:        recipeRepository,
		ingredientLoaderService: ingredientLoaderService,
		costCalculationService:  costCalculationService,
	}
}

// Execute executes the calculate recipe cost use case
func (uc *CalculateRecipeCostUseCase) Execute(recipeId string) (response.RecipeCostDto, error) {
	id := valueobject.Id{}.Of(recipeId)

	recipe, err := uc.recipeRepository.FindById(id)
	if err != nil {
		return response.RecipeCostDto{}, errors.NewRecipeNotFoundError("Recipe not found with ID: " + recipeId)
	}

	ingredientMap, err := uc.ingredientLoaderService.LoadIngredients(recipe.Ingredients())
	if err != nil {
		return response.RecipeCostDto{}, err
	}

	recipeCost, err := uc.costCalculationService.CalculateCost(recipe, ingredientMap)
	if err != nil {
		return response.RecipeCostDto{}, err
	}

	return response.FromRecipeCost(recipeCost), nil
}
